glrcfg/src/runner/executors/docker.rs

```diff
@@ -10,35 +9,282 @@ macro_rules! stringvec {
+#[derive(Debug, Serialize)]
+pub struct Services {
+    pub name: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub alias: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub entrypoint: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub command: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub environment: Option<Vec<String>>,
+}
+
+/// The image pull policy: `never`, `if-not-present` or `always` (default).
+///
+/// View details in the [pull policies documentation](https://docs.gitlab.com/runner/executors/docker.html#configure-how-runners-pull-images).
+/// You can also add [multiple pull policies](https://docs.gitlab.com/runner/executors/docker.html#set-multiple-pull-policies), [retry a failed pull](https://docs.gitlab.com/runner/executors/docker.html#retry-a-failed-pull), or [restrict pull policies](https://docs.gitlab.com/runner/executors/docker.html#allow-docker-pull-policies).
+#[derive(Debug)]
+pub enum PullPolicy {
+    Single(Option<String>),
+    Multiple(Vec<String>),
+}
+
+impl Serialize for PullPolicy {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            PullPolicy::Single(Some(s)) => serializer.serialize_some(s),
+            PullPolicy::Single(None) => serializer.serialize_none(),
+            PullPolicy::Multiple(vec) => vec.serialize(serializer),
+        }
+    }
+}
+
+fn is_pull_policy_empty(pull_policy: &PullPolicy) -> bool {
+    match pull_policy {
+        PullPolicy::Single(None) => true,
+        PullPolicy::Multiple(vec) => vec.is_empty(),
+        _ => false,
+    }
+}
+
+#[derive(Debug)]
+pub struct SecurityOpt {
+    pub key: String,
+    pub value: String,
+}
+
+impl Serialize for SecurityOpt {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let s = format!("{}:{}", self.key, self.value);
+        serializer.serialize_str(&s)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use pretty_assertions::assert_eq;
+    use serde_json;
+
+    use super::{PullPolicy, SecurityOpt};
+
+    #[test]
+    fn security_opt_serialization() {
+        let opt = SecurityOpt {
+            key: "warbl".to_string(),
+            value: "garbl".to_string(),
+        };
+
+        let serialized = serde_json::to_string(&opt).unwrap();
+        assert_eq!(serialized, "\"warbl:garbl\"");
+    }
+
+    #[test]
+    fn test_pull_policy_single_some() {
+        let policy = PullPolicy::Single(Some("single_policy".to_string()));
+        let serialized = serde_json::to_string(&policy).unwrap();
+        assert_eq!(serialized, "\"single_policy\"");
+    }
+
+    #[test]
+    fn test_pull_policy_single_none() {
+        let policy = PullPolicy::Single(None);
+        let serialized = serde_json::to_string(&policy).unwrap();
+        assert_eq!(serialized, "null");
+    }
```

Comment:
```suggestion
```

no longer needed with the introduced changes; that aside, we wouldn't _want_ a `None` pull policy to serialize to a `null` value in JSON, would we?